lib/common/ipc_controld.c

```diff
@@ -168,6 +168,14 @@ dispatch(pcmk_ipc_api_t *api, xmlNode *reply)
         pcmk_controld_reply_unknown, NULL, NULL,
     };
 
+    /* If we got an ACK, return true so the caller knows to expect more responses
+     * from the IPC server.  We do this before decrementing replies_expected because
+     * ACKs are not going to be included in that value.
+     */
+    if (pcmk__str_eq(crm_element_name(reply), "ack", pcmk__str_none)) {
+        return true;
+    }
```

Comment:
It's possible some existing commands do expect an ack, possibly with important information attached. Unfortunately the controller does not have IPC regression tests so we just need to test every available command (crmadmin, crm_node, and crm_resource have options that cover it).

Especially since the crmadmin etc commands currently work from remote nodes, there must be something handling the acks already somewhere. It could be a sync vs async thing.

Aside: The fencer and executor have standalone modes for testing. The attribute manager, CIB manager, and controller don't, so they're a weak area in our testing (the CTS lab is the only thing that exercises any of that). The scheduler's normal mode would be sufficient for ipc testing but we currently don't test the daemon ipc, just the library code (of course the ipc portion is trivial and almost never changes).

Reply:
> Especially since the crmadmin etc commands currently work from remote nodes, there must be something handling the acks already somewhere. It could be a sync vs async thing.

When I switched crm_attribute to connect with a main loop instead of sync, everything worked without patching.  So yeah, I think there's definitely a sync vs. async thing going on.

crm_node uses pcmk_ipc_dispatch_sync in controller_remove_node, but it doesn't check for a reply so it doesn't matter whether there's an ACK first or not.

crmadmin pawns all its IPC calls off to libpacemaker (pcmk__controller_status, pcmk__pacemakerd_status, etc.) and those all use a main loop.  crm_resource likewise also uses a main loop.  I'm not sure what difference between the two styles is causing the main loop approach to work.

Reply:
Hm, note that the dispatch methods in ipc_pacemakerd.c and ipc_schedulerd.c also check for an ack right away in a very similar block to what I've added here.

Reply:
The sync vs async difference appears to be due to pcmk__send_ipc_request() setting the crm_ipc_client_response flag for sync. I suspect the ack is from crmd_remote_proxy_cb() when that flag is set. I think this change won't adversely affect anything, but I'm not 100% sure. If all the regression tests (including fencer/executor) and CTS lab don't show problems, it should be OK.

Reply:
All regression tests pass.  Now to reset VMs here for running CTS lab.

Reply:
I haven't checked every single line, but it looks like the only test failures and BadNews that occurs with these patches also occurs on main.

Reply:
Looking at it again, daemons/controld/controld_control.c:dispatch_controller_ipc() sends an ack for every request that comes over local IPC. I believe the ack for remote connections in daemons/controld/controld_execd_state.c:crmd_remote_proxy_cb() is meant to do the same for remote clients (it should probably use pcmk__ipc_create_ack_as() to be more similar, though).

The question is whether the ack is already ignored somehow for the expected reply count. I think cleanup/refresh is the only thing that cares about the count, so it should be possible to tell by tracing their IPC. The important thing is that it waits for the exact number of replies -- too few and the remaining failures might not get cleaned, too many and it will hang.

Reply:
Having traced the IPC for the crm_resource -C case, it looks like response handling is correct.